### ISSUE TITLE
How to do cross server cache deletes

### ISSUE THREAD

**Issue by bep** [alexandres/poormanscdn#8]
Say I setup `example.com` to round robin to host `A` and `B`. I assume that I now, to make sure that I purge all servers, have to make several `DELETE` requests to get to them all. And note that I don't expect that this software already has a solution to this, just wanted to start a discussion.

**Comment by alexandres** [comment]
Yes, you would need to call each server for a given site. Doing with TLS is a bit trickier. You need to call `A` and `B` directly but let them know that you are contacting on behalf of `example.com`. Without TLS it's as simple as setting the `Host` header to `example.com`. With TLS you need to use [https://en.wikipedia.org/wiki/Server_Name_Indication]. The question is then how to specify an SNI servername when calling an IP in an HTTP lib that supports SNI. For example, Python [http://docs.python-requests.org/en/master/]) supports SNI, but I don't see any way to do this (call `A` specifying servername `example.com`).

Would foregoing TLS for the DELETE requests be out of the question?

As a friendly way of doing deploys, you could perhaps include in a site's repo a Cdnfile that has the IPs of all the poormancdn server and upon commit make the non-TLS DELETE calls with Host headers set to each of them.

**Comment by bep** [comment]
Maintaining a `cdnfile` per site is kind of introducing the same problem I want to solve by this: Simplify setup and maintenance.

I have an idea that might work:

I just tested by adding two IP addresses to one A record and doing a `dig mydomain.com`  -- and it lists both IP addresses in the answer section. So, if I could build that logic into my deployment routine, I only need to know the domain of the site in question.

**Comment by alexandres** [comment]
That's a better idea. Please commit the script if you implement it. 

Perhaps Cdnhost could contain only a single line indicating the hostname pointing to the poormancdn servers. Your script could then read this file and execute the `A` record lookups and DELETEs. 

You do need to find an HTTP lib that supports SNI (see my previous comment).

**Comment by bep** [comment]
This hacky snippet does what I need:

```go
package main

import (
	"context"
	"log"
	"net"
	"net/http"
	"strings"
	"time"
)

func main() {

	domain := "example.com"
	nodes, err := net.LookupHost(domain)
	if err != nil {
		log.Fatal(err)
	}

	dialer := &net.Dialer{
		Timeout:   30 * time.Second,
		KeepAlive: 30 * time.Second,
		DualStack: true,
	}

	url := "https://" + domain

	for _, node := range nodes {
		// This could be less hacky with https://github.com/golang/go/issues/12503
		transport := &http.Transport{
			DialContext: func(ctx context.Context, network, addr string) (net.Conn, error) {
				addr = node + ":" + strings.Split(addr, ":")[1]
				return dialer.DialContext(ctx, network, addr)
			},
		}

		c := &http.Client{Timeout: 10 * time.Second, Transport: transport}

		req, err := http.NewRequest("HEAD", url, nil)
		if err != nil {
			log.Fatal(err)
		}

		resp, err := c.Do(req)
		if err != nil {
			log.Fatal(err)
		}
		resp.Body.Close()
	}

}

```

**Comment by alexandres** [comment]
@bep Let me know if it works and if so please create PR. Just have to think of a good dir name to place this within the repo.

**Comment by bep** [comment]
> Let me know if it works 

It works, but it is so ugly that I thought I should let it live in some dark corner on my own repo for a while, but I'll reopen this to track it. I'm sure this is an issue other than me will encounter.
